sdk/helper/random/string_generator.go

```diff
@@ -0,0 +1,311 @@
+	return string(candidate), nil
+}
+
+const (
+	// maxCharsetLen is the maximum length a charset is allowed to be when generating a candidate string.
+	// This is the total number of numbers available for selecting an index out of the charset slice.
+	maxCharsetLen = 256
+)
+
+// randomRunes creates a random string based on the provided charset. The charset is limited to 255 characters, but
+// could be expanded if needed. Expanding the maximum charset size will decrease performance because it will need to
+// combine bytes into a larger integer using binary.BigEndian.Uint16() function.
+func randomRunes(rng io.Reader, charset []rune, length int) (candidate []rune, err error) {
+	if len(charset) == 0 {
+		return nil, fmt.Errorf("no charset specified")
+	}
+	if len(charset) > maxCharsetLen {
+		return nil, fmt.Errorf("charset is too long: limited to %d characters", math.MaxUint8)
+	}
+	if length <= 0 {
+		return nil, fmt.Errorf("unable to generate a zero or negative length runeset")
+	}
+
+	// This can't always select indexes from [0-maxCharsetLen) because it could introduce bias to the character selection.
+	// For instance, if the length of the charset is [a-zA-Z0-9-] (length of 63):
+	// RNG ranges: [0-62][63-125][126-188][189-251] will equally select from the entirety of the charset. However,
+	// the RNG values [252-255] will select the first 4 characters of the charset while ignoring the remaining 59.
+	// This results in a bias towards the front of the charset.
+	//
+	// To avoid this, we determine the largest integer multiplier of the charset length that is <= maxCharsetLen
+	// For instance, if the maxCharsetLen is 256 (the size of one byte) and the charset is length 63, the multiplier
+	// equals 4:
+	//   256/63 => 4.06
+	//   Trunc(4.06) => 4
+	// Multiply by the charset length
+	// Subtract 1 to account for 0-based counting and you get the max index value: 251
+	maxAllowedRNGValue := (maxCharsetLen/len(charset))*len(charset) - 1
+
+	// rngBufferMultiplier increases the size of the RNG buffer to account for lost
+	// indexes due to the maxAllowedRNGValue
+	rngBufferMultiplier := 1.0
+
+	// Don't set a multiplier if we are able to use the entire range of indexes
+	if maxAllowedRNGValue < maxCharsetLen {
+		// Anything more complicated than an arbitrary percentage appears to have little practical performance benefit
+		rngBufferMultiplier = 1.5
+	}
+
+	if rng == nil {
+		rng = rand.Reader
+	}
+
+	charsetLen := byte(len(charset))
+
+	runes := make([]rune, 0, length)
+
+	for len(runes) < length {
+		// Generate a bunch of indexes
+		data := make([]byte, int(float64(length)*rngBufferMultiplier))
+		numBytes, err := rng.Read(data)
+		if err != nil {
+			return nil, err
+		}
+
+		// Append characters until either we're out of indexes or the length is long enough
+		for i := 0; i < numBytes; i++ {
+			if data[i] > byte(maxAllowedRNGValue) {
+				continue
+			}
+
+			// Be careful about byte conversions - you don't want 256 to wrap to 0
```

Comment:
If the charset length is 256 wouldn’t the 256 index be out of range?

Reply:
The warning here is regarding accidentally converting an int to a byte where the int is 256, resulting in `byte(256)` being converted to `0`. I did this a couple of times and felt it warranted a warning to be careful about it.

I do have a unit test that checks the generation with a 256 length charset as well (which I just realized is off by one, so that is fixed in the next commit).

Reply:
Okay, i think this comment confused me. Based on it i was thinking index below could be 256, but that's not the case both because that would overflow `byte` and because the maxAllowedRNGValue can only be a max of 255. Maybe make this comment more clear or remove it if it's no longer relevant?

Reply:
:+1: